### PR TITLE
Table array access in TreeSequence

### PR DESCRIPTION
**Pull request by jeromekelleher**
Add array access to TreeSequence and solve some performance issues/memory leaks. See below for more details.

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Before using the standard ``ts.trees()`` approach:

timedesc-leak-before-trees

Before using the sequential ``ts.at(j)`` approach:
timedesc-leak-before-at

This branch:

timedesc-leak-after

The difference is pretty stark - 60M vs around 650M. It's interesting that it doesn't seem like this is a real "leak" per-se, because the memory does hit a high-water mark. It's an awful lot more though, and maybe we're having to run garbage collection or something to get it back?

**Comment by jeromekelleher** [comment]
Fixes #1916:

Benchmark:
```python
import msprime
import time


before = time.perf_counter()
ts = msprime.sim_ancestry(
    10000,
    sequence_length=1000000,
    population_size=10_000,
    recombination_rate=1e-8,
    random_seed=1234,
)
duration = time.perf_counter() - before
print(f"Simulation of {ts.num_trees} trees done after {duration:.2f} seconds")

for _ in range(1000):
    samples = ts.samples(time=0)
    # print(samples)
```
Before:
samples-mem-before

This branch:

samples-mem-after

**Comment by jeromekelleher** [comment]
Performance improvement of ``ts.site(position=x)``. Benchmark:
```python
import msprime
import time


before = time.perf_counter()
ts = msprime.sim_ancestry(
    10000,
    sequence_length=1000000,
    population_size=10_000,
    recombination_rate=1e-8,
    random_seed=1234,
)
ts = msprime.sim_mutations(ts, rate=1e-7, random_seed=1)
duration = time.perf_counter() - before
print(f"Simulation of {ts.num_trees} trees done after {duration:.2f} seconds")

for site in ts.sites():
    s = ts.site(position=site.position)
    if site.id >= 1000:
        break
```

Before:
site-search-before

After:
site-search-after

**Comment by jeromekelleher** [comment]
Performance for using ``ts.variants(left=x)``:

```python
import msprime
import time


before = time.perf_counter()
ts = msprime.sim_ancestry(
    10000,
    sequence_length=1000000,
    population_size=10_000,
    recombination_rate=1e-8,
    random_seed=1234,
)
ts = msprime.sim_mutations(ts, rate=1e-7, random_seed=1)
duration = time.perf_counter() - before
print(f"Simulation of {ts.num_trees} trees done after {duration:.2f} seconds")

for _ in range(10):
    for var in ts.variants(left=1):
        pass
```
Before:
varstart-before

After:
varstart-after

**Comment by jeromekelleher** [comment]
Running edge_diffs:
```python
import msprime
import time


before = time.perf_counter()
ts = msprime.sim_ancestry(
    10000,
    sequence_length=1000000,
    population_size=10_000,
    recombination_rate=1e-8,
    random_seed=1234,
)
# ts = msprime.sim_mutations(ts, rate=1e-7, random_seed=1)
duration = time.perf_counter() - before
print(f"Simulation of {ts.num_trees} trees done after {duration:.2f} seconds")

before = time.perf_counter()
for _ in range(100):
    for _ in ts.edge_diffs():
        pass
duration = time.perf_counter() - before

print(f"Ran edge diffs in {duration:.2f} seconds")
```
Before
edge_diffs_before

After:

edge_diffs_before
:

**Comment by jeromekelleher** [comment]
This is ready for review. The basic idea is to provide array access via the TreeSequence object in lieu of the "proper" solution of making TreeSequence.tables a read-only view (which is hard, engineering wise). On the way I've moved all internal usage of ``self.tables`` as they were all leading to significant performance problems (at best) and sometimes major memory leaks (#2428). I don't understand why some of these were leaking memory and others aren't, but we can follow that up later - at least library functions shouldn't leak and be efficient.

Since we've got some memory leaks in the released version we should probably get this out ASAP.

File: python/tskit/trees.py [hunk 1]
```diff
@@ -3860,6 +3860,160 @@ class TreeSequence:
     the :meth:`.variants` method iterates over all sites and their genotypes.
     """
 
+    individuals_flags: np.array
+    """
+    Efficient access to the ``flags`` column in the
+    :ref:`sec_individual_table_definition` as a numpy array (dtype=np.uint32).
+    Equivalent to ``ts.tables.individuals.flags``.
```
Comment (petrelharp) [inline python/tskit/trees.py:3867]: "Equivalent to" makes it sound like there's no difference (the "Efficient" above notwithstanding). Maybe "Obtains the same values as `xxx` but without copying the entire tables."? Or, maybe it's okay?

**Review by petrelharp — APPROVED** [review]
NICE!!! Looks good to me - just a docs suggestion/comment.

File: python/tskit/trees.py (hunk 1 above)
Comment (jeromekelleher) [reply python/tskit/trees.py]: I've added an extrac note about this

**Comment by jeromekelleher** [comment]
Marking this one for merging as I need it to work with the 1.4M sample tree sequences on Zenodo (and there's another bug that needs resolving to get a VCF out of them). Please do follow up with issues for any review problems you spot @benjeffery @hyanwong

**Comment by hyanwong** [comment]
Sorry for the slow review. LGTM - this is a great improvement. My only query would be if / how we intend to maintain this in the long term, once we address #760. Will we then deprecate these attributes, and simply have them as permanently maintained (and undocumented?) equivalents to `ts.tables.nodes.time` and friends?

**Comment by jeromekelleher** [comment]
> Will we then deprecate these attributes, and simply have them as permanently maintained (and undocumented?) equivalents to ts.tables.nodes.time and friends?

I think we can keep them permanently as part of the public API. There is an advantage of fewer attribute lookups in ``ts.node_flags`` to ``ts.tables.nodes.flags`` as well as a bit less typing, so no harm in keeping them I think.

**Review by benjeffery — COMMENTED** [review]
Been through this closely and looks good!

File: python/tests/test_highlevel.py [hunk 2]
```diff
@@ -2591,6 +2592,81 @@ def test_individual_properties(self, n):
...
+        with pytest.raises(AttributeError, match="delete attribute"):
+            delattr(ts, array)
+        with pytest.raises(ValueError, match="read-only"):
+            a[:] = 1
+
+    def test_arrays_equal_to_tables(self, ts_fixture):
+        ts = ts_fixture
+        tables = ts.tables
+
+        assert_array_equal(ts.individuals_flags, tables.individuals.flags)
```
[the remark this answers is not in the document]
Comment (jeromekelleher) [reply python/tests/test_highlevel.py:2637]: I thought this was simpler and easier to follow. There would have been some obscure string parsing and getattring in order to get the corresponding table array, and this seemed more ... obvious.

**Pull request merged**